### ISSUE TITLE
32-bit Build Failure with CCPP

### ISSUE THREAD

**Issue by stefan-vaisala** [ufs-community/ufs-weather-model#328]
## Description
On version tag GFS.v16.0.14, following all necessary versions for all submodules, the build fails using Intel compilers with the 32 bit flag enabled for dycore when compiled with CCPP. The failure is due to type disagreements with CCPP within "atmos_cubed_sphere/model/fv_dynamics.F90". 

Upon further examination, it looks like the CCPP CMakeLists.txt file looks for the CMake variable "DYN32" to be set in order to properly set for a dycore 32-bit build. This CMake variable is left undefined by the overarching CMake process as executed by "build.sh", and thus, the CCPP builds with not setting -DOVERLOAD_R4, and the typing mistmatch.

The fix is to define the "DYN32" variable within the 32BIT if-block within Intel.cmake (https://github.com/ufs-community/ufs-weather-model/blob/GFS.v16.0.14/cmake/Intel.cmake#L6). 

E.g.,
```if(32BIT)
    set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -i4 -real-size 32")
    add_definitions(-DOVERLOAD_R4)
    add_definitions(-DOVERLOAD_R8)
    set(DYN32 "ON")
```
This above change allows the build to proceed successfully - though I have not tested if the model runs yet. There are similar CCPP CMake variables that are not properly defined by the overarching CMake build process, too, such as ESMF_LIB_DIR. The ESMF_LIB_DIR can be fixed by adding "set(ESMF_LIB_DIR "${ESMF_LIBSDIR}")" to the used CMake configuration file (e.g., configure_linux.intel.cmake). 

- Stefan

**Comment by junwang-noaa** [comment]
Tag GFS.v16.0.14  only supports IPD and gnu make, the code was diverged
from develop branch since Feb, 2020. Please go to the ufs-weather-model
develop branch for CCPP build capability and CCPP code equivalent  to
GFS.v16.0.14 IPD physics.

**Comment by stefan-vaisala** [comment]
@junwang-noaa Thanks for the info. What is the proper "build" procedures for Tag GFS.v16.0.14 using build.sh then? I assume its disabling CCPP somehow, but not sure.

Looks like it could just be changing "CMAKE_FLAGS+=" -DCCPP=ON -DSTATIC=ON -DSUITES=${CCPP_SUITES} -DNETCDF_DIR=${NETCDF}"" in build.sh to "CMAKE_FLAGS+=" -DCCPP=OFF"".

**Comment by climbfuji** [comment]
I think the way to build the GFS.v16.0.14 tag is either using `tests/compile.sh` or running the global workflow (which builds it in the same way). The `compile.sh` in this tag is different from develop and should take arguments like that:
```
cd tests
./compile.sh $PWD/../FV3 MACHINE.COMPILER 'CCPP=N 32BIT=Y' [LABEL] [clean_before] [clean_after]
```
The question is whether your machine and compiler are supported (mostly NOAA HPC SYSTEMS, but generic `linux.gnu` and `macosx.gnu` may also work).

`build.sh` with `CCPP=OFF` may work, too (and drop the suites argument).

Bottom line is that this tag isn't supported to the community and community platforms. The develop branch or the public releases are better choices.

**Comment by stefan-vaisala** [comment]
build.sh with "CMAKE_FLAGS+=" -DCCPP=OFF"" works just as expected - skips the CCPP builds and the model still compiles (incl IPD). Thanks for the other compile method - might give that a go, too. Only reason I went the build.sh route was that it has a linux.intel build configuration I could leverage, while for the test/compile.sh method, I'll have to create the linux.intel configuration (which shouldnt be too hard to emulate from existing intel configurations).
